File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int countPairs(TreeNode* root, int distance) {
        vector<string> leafs;
        queue<pair<TreeNode*, string>> q;
        q.push(make_pair(root, ""));
        
        pair<TreeNode*, string> p;
        TreeNode* node;
        string dist;
        while(!q.empty()) {
            p = q.front();
            node = p.first;
            dist = p.second;
            q.pop();
            
            if(node->left == NULL && node->right == NULL) { //리프노드
                leafs.push_back(dist);
            } else {
                if(node->left != NULL) q.push(make_pair(node->left, dist + "L"));
                if(node->right != NULL) q.push(make_pair(node->right, dist + "R"));
            }
        }
        
        int answer = 0, k;
        
        for(int i = 0; i < leafs.size(); i++) {
            for(int j = i + 1; j < leafs.size(); j++) {
                k = 0;
                while(leafs[i][k] == leafs[j][k]) {
                    k++;
                }
                
                if(leafs[i].size() + leafs[j].size() - k * 2 <= distance) {
                    answer++;
                }
            }
        }
        
        return answer;
    }
};

File: 1530-number-of-good-leaf-nodes-pairs/1530-number-of-good-leaf-nodes-pairs.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

import queue

class Solution:
    def countPairs(self, root: TreeNode, distance: int) -> int:
        leafs = []
        
        q = queue.Queue()
        q.put([root, ""])
        
        node, dist = None, None
        while q.empty() == False:
            node, dist = q.get()
            if node.left == None and node.right == None:
                leafs.append(dist)
            if node.left != None:
                q.put([node.left, dist + "L"])
            if node.right != None:
                q.put([node.right, dist + "R"])
        
        answer = 0
        for i, leaf1 in enumerate(leafs):
            for leaf2 in leafs[i + 1:]:
                k = 0
                while leaf1[k] == leaf2[k]:
                    k += 1
                
                if len(leaf1) + len(leaf2) - k * 2 <= distance:
                    answer += 1
        
        
        return answer